core/src/layers/5-disk/sworndisk.rs

```diff
@@ -0,0 +1,812 @@
+        Ok(new_self)
+    }
+
+    /// Opens the `SwornDisk` on the given disk, with the root encryption key.
+    pub fn open(disk: D, root_key: Key) -> Result<Self> {
+        let data_disk = Self::subdisk_for_data(&disk)?;
+        let lsm_tree_disk = Self::subdisk_for_logical_block_table(&disk)?;
+
+        let tx_log_store = Arc::new(TxLogStore::recover(lsm_tree_disk, root_key)?);
+        let block_validity_table = Arc::new(AllocTable::recover(
+            NonZeroUsize::new(data_disk.nblocks()).unwrap(),
+            &tx_log_store,
+        )?);
+        let listener_factory = Arc::new(TxLsmTreeListenerFactory::new(
+            tx_log_store.clone(),
+            block_validity_table.clone(),
+        ));
+
+        let logical_block_table = {
+            let table = block_validity_table.clone();
+            let on_drop_record_in_memtable = move |record: &dyn AsKV<RecordKey, RecordValue>| {
+                // Deallocate the host block while the corresponding record is dropped in `MemTable`
+                table.set_deallocated(record.value().hba);
+            };
+            let master_sync_id = MasterSyncId::load(Self::subdisk_for_master_sync_id(&disk)?)?;
+            TxLsmTree::recover(
+                tx_log_store.clone(),
+                listener_factory,
+                Some(Arc::new(on_drop_record_in_memtable)),
+                master_sync_id,
+            )?
+        };
+
+        let opened_self = Self {
+            inner: Arc::new(DiskInner {
+                bio_req_queue: BioReqQueue::new(),
+                logical_block_table,
+                user_data_disk: data_disk,
+                block_validity_table,
+                data_buf: DataBuf::new(DATA_BUF_CAP),
+                tx_log_store,
+                root_key,
+                is_dropped: AtomicBool::new(false),
+                sync_region: Mutex::new(()),
+            }),
+        };
+
+        // opened_self.spawn_bio_req_handler();
+
+        #[cfg(not(feature = "linux"))]
+        debug!("[SwornDisk] Opened successfully!");
+        Ok(opened_self)
+    }
+
+    /// Returns the total number of blocks in the device.
+    pub fn total_blocks(&self) -> usize {
+        self.inner.user_data_disk.nblocks()
+    }
+
+    /// Spawn a task to continuously handle any queued block I/O requests.
+    // Currently not in-use.
+    fn spawn_bio_req_handler(&self) {
+        let inner = self.inner.clone();
+        let _ = crate::os::spawn(move || {
+            inner.handle_bio_reqs_looped();
+        });
+    }
+
+    fn subdisk_for_data(disk: &D) -> Result<D> {
+        disk.subset(0..disk.nblocks() * 15 / 16) // TBD
+    }
+
+    fn subdisk_for_logical_block_table(disk: &D) -> Result<D> {
+        disk.subset(disk.nblocks() * 15 / 16..disk.nblocks() - 1) // TBD
+    }
+
+    fn subdisk_for_master_sync_id(disk: &D) -> Result<D> {
+        // Use last block for storing master sync ID for now
+        disk.subset(disk.nblocks() - 1..disk.nblocks()) // TBD
```

Comment:
Refactored as follow:
```rust
/// Master sync ID to help `TxLsmTree` achieve sync awareness.
pub struct MasterSyncId {
    id: AtomicU64,
    store: Arc<dyn SyncIdStore>,
}

/// A trusted store that stores the master sync ID.
pub trait SyncIdStore {
    /// Read the current master sync ID from the store.
    fn read(&self) -> Result<SyncId>;

    /// Write the given master sync ID to the store.
    fn write(&self, id: SyncId) -> Result<()>;
}

/// Currently we use untrusted store `D: BlockSet` to store the
/// master sync ID thus the security guarantee is downgraded accordingly.
///
/// TODO: Use external trusted store to store the master sync ID.
struct SyncIdDiskStore<D> {
    disk: D,
}

impl<D: BlockSet + 'static> SyncIdStore for SyncIdDiskStore<D> {
    /*  */
}
```
Still need a default untrusted disk store for now since the master sync ID must be persisted somewhere to help `SwornDisk` recover.